### PR TITLE
SAGA: Add keymapper support

### PR DESCRIPTION
**Pull request by NabeelShabbir**

### CODE REVIEW

**Comment by bluegr** [comment]
Thanks for your work!
These are a lot of changes fir a single commit, can you please split your changes into smaller commits?

**Comment by NabeelShabbir** [comment]
> Thanks for your work! These are a lot of changes for a single commit, can you please split your changes into smaller commits?

Thank you, I'll take care of it.

File: engines/saga/displayinfo.h [hunk 1]
```diff
@@ -157,14 +157,14 @@ static PanelButton ITE_MainPanelButtons[] = {
...
-	{kPanelButtonVerb,		132,0,	34,16,	kVerbITEUse,'u',0,		10,11,0},
-	{kPanelButtonVerb,		132,17,	34,16,	kVerbITEGive,'g',0,	12,13,0},
+	{kPanelButtonVerb,		53,0,	34,16,	kVerbITEWalkTo,kActionWalkTo,0,	0,1,0},
+	{kPanelButtonVerb,		53,17,	34,16,	kVerbITELookAt,kActionLookAt,0,	2,3,0},
+	{kPanelButtonVerb,		53,34,	34,16,	kVerbITEPickUp,kActionPickUp,0,	4,5,0},
+	{kPanelButtonVerb,		92,0,	34,16,	kVerbITETalkTo,kActionTalkTo,0,	0,1,0},
+	{kPanelButtonVerb,		92,17,	34,16,	kVerbITEOpen,kActionOpen,0,	6,7,0},
+	{kPanelButtonVerb,		92,34,	34,16,	kVerbITEClose,kActionClose,0,	8,9,0},
+	{kPanelButtonVerb,		132,0,	34,16,	kVerbITEUse,kActionUse,0,		10,11,0},
+	{kPanelButtonVerb,		132,17,	34,16,	kVerbITEGive,kActionGive,0,	12,13,0},
```
Comment (bluegr) [inline engines/saga/displayinfo.h:167]:
```suggestion
	{kPanelButtonVerb,		 53, 0,	34,16,	kVerbITEWalkTo,	kActionWalkTo,	0,	0,1,0},
	{kPanelButtonVerb,		 53,17,	34,16,	kVerbITELookAt,	kActionLookAt,	0,	2,3,0},
	{kPanelButtonVerb,		 53,34,	34,16,	kVerbITEPickUp,	kActionPickUp,	0,	4,5,0},
	{kPanelButtonVerb,		 92, 0,	34,16,	kVerbITETalkTo,	kActionTalkTo,	0,	0,1,0},
	{kPanelButtonVerb,		 92,17,	34,16,	kVerbITEOpen,	kActionOpen,	0,	6,7,0},
	{kPanelButtonVerb,		 92,34,	34,16,	kVerbITEClose,	kActionClose,	0,	8,9,0},
	{kPanelButtonVerb,		132, 0,	34,16,	kVerbITEUse,	kActionUse,		0,	10,11,0},
	{kPanelButtonVerb,		132,17,	34,16,	kVerbITEGive,	kActionGive,	0,	12,13,0},
```

File: engines/saga/displayinfo.h [hunk 2]
```diff
@@ -190,13 +190,13 @@ static PanelButton ITE_ConversePanelButtons[] = {
...
-	{kPanelButtonOption,	13,98, 135,17,	kTextContinuePlaying,'c',0,	0,0,0}, //continue
-	{kPanelButtonOption,	164,98, 57,17,	kTextLoad,'l',0,	0,0,0}, //load
-	{kPanelButtonOption,	241,98, 57,17,	kTextSave,'s',0,	0,0,0},	//save
+	{kPanelButtonOption,	113,18, 45,17,	kTextReadingSpeed,kActionOptionReadingSpeed,0,	0,0,0}, //read speed
+	{kPanelButtonOption,	113,37, 45,17,	kTextMusic,kActionOptionMusic,0,	0,0,0}, //music
+	{kPanelButtonOption,	113,56, 45,17,	kTextSound,kActionOptionSound,0,	0,0,0}, //sound-noise
+	{kPanelButtonOption,	13,79, 135,17,	kTextQuitGame,kActionOptionQuitGame,0,	0,0,0}, //quit
+	{kPanelButtonOption,	13,98, 135,17,	kTextContinuePlaying,kActionOptionContinue,0,	0,0,0}, //continue
+	{kPanelButtonOption,	164,98, 57,17,	kTextLoad,kActionOptionLoad,0,	0,0,0}, //load
+	{kPanelButtonOption,	241,98, 57,17,	kTextSave,kActionOptionSaveGame,0,	0,0,0},	//save
```
Comment (bluegr) [inline engines/saga/displayinfo.h:199]:
```suggestion
	{kPanelButtonOption,	113,18, 45,17,	kTextReadingSpeed,		kActionOptionReadingSpeed,	0,	0,0,0}, //read speed
	{kPanelButtonOption,	113,37, 45,17,	kTextMusic,				kActionOptionMusic,			0,	0,0,0}, //music
	{kPanelButtonOption,	113,56, 45,17,	kTextSound,				kActionOptionSound,			0,	0,0,0}, //sound-noise
	{kPanelButtonOption,	13,79, 135,17,	kTextQuitGame,			kActionOptionQuitGame,		0,	0,0,0}, //quit
	{kPanelButtonOption,	13,98, 135,17,	kTextContinuePlaying,	kActionOptionContinue,		0,	0,0,0}, //continue
	{kPanelButtonOption,	164,98, 57,17,	kTextLoad,				kActionOptionLoad,			0,	0,0,0}, //load
	{kPanelButtonOption,	241,98, 57,17,	kTextSave,				kActionOptionSaveGame,		0,	0,0,0},	//save
```

File: engines/saga/displayinfo.h [hunk 3]
```diff
@@ -208,13 +208,13 @@ static PanelButton ITE_OptionPanelButtons[] = {
...
```
Comment (bluegr) [inline engines/saga/displayinfo.h:217]:
```suggestion
	{kPanelButtonOption,	113,18, 45,17,	kTextReadingSpeed,		kActionOptionReadingSpeed,	0,	0,0,0}, //read speed
	{kPanelButtonOption,	113,37, 45,17,	kTextMusic,				kActionOptionMusic,			0,	0,0,0}, //music
	{kPanelButtonOption,	113,56, 45,17,	kTextSound,				kActionOptionSound,			0,	0,0,0}, //sound-noise
	{kPanelButtonOption,	13,79, 135,17,	kTextQuitGame,			kActionOptionQuitGame,		0,	0,0,0}, //quit
	{kPanelButtonOption,	13,98, 135,17,	kTextContinuePlaying,	kActionOptionContinue,		0,	0,0,0}, //continue
	{kPanelButtonOption,	164,98, 57,17,	kTextLoad,				kActionOptionLoad,			0,	0,0,0}, //load
	{kPanelButtonOption,	241,98, 57,17,	kTextSave,				kActionOptionSaveGame,		0,	0,0,0},	//save
```

File: engines/saga/displayinfo.h [hunk 4]
```diff
@@ -390,14 +390,14 @@ static const GameDisplayInfo ITE_DisplayInfo_ZH = {
...
-	{kPanelButtonVerb,		223,76,		114,30, kVerbIHNMGive,'g',0,	12,13,0},
-	{kPanelButtonVerb,		223,108,	114,30, kVerbIHNMPush,'p',0,	14,15,0},
+	{kPanelButtonVerb,		106,12,		114,30,	kVerbIHNMWalk,kActionWalkTo,0,	0,1,0},
+	{kPanelButtonVerb,		106,44,		114,30,	kVerbIHNMLookAt,kActionLookAt,0,	2,3,0},
+	{kPanelButtonVerb,		106,76,		114,30, kVerbIHNMTake,kActionTake,0,	4,5,0},
+	{kPanelButtonVerb,		106,108,	114,30, kVerbIHNMUse,kActionUse,0,		6,7,0},
+	{kPanelButtonVerb,		223,12,		114,30, kVerbIHNMTalkTo,kActionTalkTo,0,	8,9,0},
+	{kPanelButtonVerb,		223,44,		114,30, kVerbIHNMSwallow,kActionSwallow,0,	10,11,0},
+	{kPanelButtonVerb,		223,76,		114,30, kVerbIHNMGive,kActionGive,0,	12,13,0},
+	{kPanelButtonVerb,		223,108,	114,30, kVerbIHNMPush,kActionPush,0,	14,15,0},
```
Comment (bluegr) [inline engines/saga/displayinfo.h:400]:
```suggestion
	{kPanelButtonVerb,		106,12,		114,30,	kVerbIHNMWalk,		kActionWalkTo,	0,	 0, 1,0},
	{kPanelButtonVerb,		106,44,		114,30,	kVerbIHNMLookAt,	kActionLookAt,	0,	 2, 3,0},
	{kPanelButtonVerb,		106,76,		114,30, kVerbIHNMTake,		kActionTake,	0,	 4, 5,0},
	{kPanelButtonVerb,		106,108,	114,30, kVerbIHNMUse,		kActionUse,		0,	 6, 7,0},
	{kPanelButtonVerb,		223,12,		114,30, kVerbIHNMTalkTo,	kActionTalkTo,	0,	 8, 9,0},
	{kPanelButtonVerb,		223,44,		114,30, kVerbIHNMSwallow,	kActionSwallow,	0,	10,11,0},
	{kPanelButtonVerb,		223,76,		114,30, kVerbIHNMGive,		kActionGive,	0,	12,13,0},
	{kPanelButtonVerb,		223,108,	114,30, kVerbIHNMPush,		kActionPush,	0,	14,15,0},
```

File: engines/saga/saga.h [hunk 5]
```diff
@@ -631,8 +678,12 @@ class SagaEngine : public Engine {
 	const char *getTextString(int textStringId);
 	void getExcuseInfo(int verb, const char *&textString, int &soundResourceId);
 
+	void enableKeyMap(int mode);
+
 private:
 
+	int _currentPanelMode;
```
Comment (bluegr) [inline engines/saga/saga.h:685]: You can use `_interface->getMode()`

File: engines/saga/interface.h [hunk 6]
```diff
@@ -284,6 +284,8 @@ class Interface {
 
 	int32 getProtectHash() { return _protectHash; }
 
+	PanelModes getPanelMode() const { return (PanelModes)_panelMode; }
```
Comment (bluegr) [inline engines/saga/interface.h:287]: You can use `getMode()` (line 192, above)

File: engines/saga/input.cpp [hunk 7]
```diff
@@ -145,6 +145,17 @@ int SagaEngine::processInput() {
...
+		Common::Keymapper *keymapper = SagaEngine::getEventManager()->getKeymapper();
+		if (_interface->_textInput || _interface->_statusTextInput) {
+			keymapper->getKeymap("game-shortcuts")->setEnabled(false);
+			keymapper->getKeymap("save-panel")->setEnabled(false);
+		} else {
+			keymapper->getKeymap("game-shortcuts")->setEnabled(true);
+			keymapper->getKeymap("save-panel")->setEnabled(true);
+		}
+
+		enableKeyMap(_interface->getPanelMode());
```
Comment (bluegr) [inline engines/saga/input.cpp:158]: You can use `_interface->getMode()`

File: engines/saga/interface.h (hunk 6 above)
Comment (NabeelShabbir) [reply engines/saga/interface.h]: Fixed

File: engines/saga/input.cpp (hunk 7 above)
Comment (NabeelShabbir) [reply engines/saga/input.cpp]: Fixed

File: engines/saga/saga.h (hunk 5 above)
Comment (bluegr) [reply engines/saga/saga.h:685]: Right. Thanks for the heads up! Yeah, this is a simple way of performing this check

**Comment by bluegr** [comment]
Thanks!

**Pull request merged**